### ISSUE TITLE
[4.0][Bug?] Adding a column without TYPE does not work for Enum

### ISSUE THREAD

**Issue by genesiscz** [Laravel-Backpack/CRUD#2558]
# Bug report

### What I did

I am using $this->crud->addColumn without the type property. It is ready for that, but not for the enum, which is not being known by Doctrine. function setDoctrineTypesMapping is in place but not used there. It is being used with setFromDb though. 

### What I've already tried to fix it
Setting the mappings like this:
```
    function __construct() {
        parent::__construct();
        $this->crud->setDoctrineTypesMapping();
    }
```
and
```
 function setupDefaults() {
        $this->crud->setDoctrineTypesMapping();
    }
```

and also
```
  function setupConfigurationForCurrentOperation() {
        $this->crud->setDoctrineTypesMapping();
        parent::setupConfigurationForCurrentOperation();
    }
```
but since the setup() is being used for adding the columns and fields (because I am using robust way to set it up for all operations at once), it doesn't execute early enough.

### Here's how I managed to fix it

I have my own function to addColumnsAndFields (see below), calling `$this->crud->setDoctrineTypesMapping`

I can send PR if you want 👍

**Comment by genesiscz** [comment]
- Unfortunatelly, I did post the issue without review and I can't change the title, so sorry for that
- Is there a better way to mix columns and fields definition without having to write their definitions twice, sometimes three times? I went with this approach

    function setup() {
       $data = [
            "store.name" => [
                "label" => __("Store"),
                "name" => "store_id",
                "column" => [
                    "key" => "name",
                    "type" => "closure",
                    "entity" => "store",
                    "attribute" => "name",
                    "function" => fn($row) => "<a href='".route("stores.show", ["id" => $row->store->id])."'>".$row->store->name."</a>",
                ],
                "field" => [
                      "type" => "select",
                      "options" => [....]
                ]
   
        ];

        $this->addColumnsAndFields($data);
    }

    function addColumnsAndFields($data) {
        foreach($data as $index => $columnData) {
            $column = $columnData;
            $field = $columnData;
            if (isset($columnData["column"])) {
                $column = array_merge($column, $columnData["column"]);
            }

            if (isset($columnData["column"])) {
                $field = array_merge($field, $columnData["field"]);
            }
            $this->crud->addColumn($column);
            $this->crud->addField($field);
        }
    }

**Comment by tabacitu** [comment]
@genesiscz I'm not sure I understand the issue. 

Let me rephrase - can you please let me know if I got it right?

- you've done ```$this->crud->addColumn('something')``` for a column that's an ENUM in the DB, but it triggers an error because ```setDoctrineTypesMapping()``` is only called if the Type is explicitly defined.

Did I get this right?

**Comment by tabacitu** [comment]
> Is there a better way to mix columns and fields definition without having to write their definitions twice, sometimes three times? 

I do something similar sometimes. But I prefer to define things inside each operation's setup method. That way:
- I don't define "columns" inside the create/update forms
- I don't define "fields" inside the list/show operation

What I basically do sometimes is:
```php
public $fieldsOrColumns;

public function setup() {
   // ..
   $this->fieldsOrColumns = [
        'name',
        [
            'name' => 'description',
            'label' => 'Description',
	    'type' => 'textarea'
	],
	[
            'name' => 'bio',
	    'label' => 'Bio',
            'type' => 'textarea'
        ]
   ]
}

public function setupListOperation() {
	CRUD::addColumns($this->fieldsOrColumns);
}

public function setupCreateOperation() {
	CRUD::addFields($this->fieldsOrColumns);
}
```

But the truth of the matter is - the Columns definition is SIMILAR to Fields most of the times, but it's very rarely IDENTICAL. So in 99% of cases I just write it once for each operation. It makes it easier to change things just for just open operation, when you need to. And it makes it clear to you and whatever developer comes after you what fields/columns are inside each operation.

-----

In 4.1 (coming in a few days/weeks) we have a few things lined up, that will improve how we can define the columns and fields. For example https://github.com/Laravel-Backpack/CRUD/pull/2513 will allow us to do this for the above:

```php
public function setupListOperation() {
	CRUD::column('name');
	CRUD::column('description');
	CRUD::column('bio');
}

public function setupCreateOperation() {
	CRUD::field('name');
	CRUD::field('description')->type('textarea');
	CRUD::field('bio')->type('ckeditor');
}
```

Which doesn't help AT ALL in not having to write it twice. But it does help to reduce the number of lines needed to define a column, since most developers are ok with chaining methods on one line, but not ok with having an associative array squished into one line.

**Comment by tabacitu** [comment]
Oh @genesiscz I've just thought of something. If you're comfortable with having Columns and Fields even in operations where the Fields/Columns are not used, something else you could do is:

```php
public function setup() {
   // ..
   CRUD::addColumns([
        'name',
        [
            'name' => 'description',
            'label' => 'Description',
	    'type' => 'textarea'
	],
	[
            'name' => 'bio',
	    'label' => 'Bio',
            'type' => 'textarea'
        ]
   ]);
}

public function setupListOperation() {
	// already defined, nothing to do
}

public function setupCreateOperation() {
        // over here, we can just define all the columns as fields
	CRUD::addFields($this->crud->columns);
}
```

Again, it's not something I would do myself. Because I found in most cases the columns/fields definition is not exactly the same. But that should be possible, and it's arguably a little cleaner than your initial solution.

**Comment by genesiscz** [comment]
> you've done $this->crud->addColumn('something') for a column that's an ENUM in the DB, but it triggers an error because setDoctrineTypesMapping() is only called if the Type is explicitly defined.

And in setFromDb, yeah.

> Again, it's not something I would do myself. Because I found in most cases the columns/fields definition is not exactly the same. But that should be possible, and it's arguably a little cleaner than your initial solution.

That's the reason i wanted to define the name/label once, but type and possibly more parameters dynamically. 

When I tried ur tip from the last post, the textarea shown up in the table as blank but as a textarea in an edit. That's bad. 

I tried to use this to show date in the table but textarea in the edit. Works perfectly and you see what attributes are used for what. 

```
       $this->addColumnsAndFields([
            "created_at" => [
                "name" => "created_at",
                "label" => "created at",
                "column" => [
                    "type" => "date",
                ],
                "field" => [
                    "type" => "textarea"
                ]
            ]
        ]);
```

Also, I am not sure if I missed that feature completelly or it just doesn't exist, but I wanted to elegantly convert the foreign key relation to a href with route to the show method of the underlying entity. 

I somehow rewritten the setFromDb to automatically convert all foreign keys to link to the show operation of that foreign key's entity. It even works with edit, where I just mark the field as select2. I believe more issues are yet to come but I wanted to play around with it a little.

I understand you can't change this method a lot since I read some of the discussions here. Even if it can't be a functionality of this method, it would be nice to see a column type that would automatically return a link without needing to creating them manually.

I also understand I am writing this wall of text to a bug report. I would create a new thread as an idea, but I am not sure it's not a crazy one.

**Comment by tabacitu** [comment]
Got it! We'll investigate and reply here about the ENUM - if something can be done about that.

But in principle, I'm not sure Backpack will ever _officially_ support defining fields&columns together, then modifying them separately for each operation, to "finish them off". I think it makes the definition process more complicated than it needs to be - more difficult to write and more difficult to read. Granted, your ```addColumnsAndFields()``` solution above is a good one if you really _really_ want to do this, but... I think most people will find this confusing. We've specifically tried to move away from defining multiple operations together:
- in v3 we recommended that people define everything in ```setup()```; everything got defined for all operations; every operation was then bloated with features they didn't use; it was difficult to define something for one operation but not another one;
- in v4 we learnt our lesson, people can _still_ define things in ```setup()```, but we recommend they configure their operations inside ```setupXxxOperation()``` methods; that way there's no way to confuse what happens for each operation; you can still use ```setFromDb()``` for both columns and fields, which would basically set all columns and fields, then modify them slightly;

Providing a way for people to define both columns and fields in one go, when NO operation uses both columns and fields, sounds like a step back - at least in my opinion. Then again, I've been wrong before. We'll leave this open and see what other people think about it 😉 But please don't expect we'll implement this anytime soon, I doubt it.

Sorry for the downvote 😄 
Cheers!

**Comment by tabacitu** [comment]
@genesiscz good news - I believe #2974 has fixed this problem, by always doing `setDoctrineTypesMapping()` before guessing field attributes. Let me know if it doesn't do it for you and we'll reopen.

Cheers!
